### PR TITLE
6724 - Alteration UI - Added component validation to CompanyProvisions and ResolutionDates

### PR DESCRIPTION
**Pull request by andrepestana-aot**
*Issue #:* /bcgov/entity#6274

*Description of changes:*

- Added component validation to CompanyProvisions and ResolutionDates
- Added tests

**This is part of 6274 regarding the component validations when the user clicks on "Review and Certify" only.**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/components/Articles/Articles.vue
Comment (severinbeauvais) [inline src/components/Articles/Articles.vue:56]: This is the same name as the company provisions component event. While this may not be causing a conflict, it may result in confusion.
How about renaming the event `isEditMode` (same as some of the other components)?

Comment (andrepestana-aot) [reply src/components/Articles/Articles.vue:56]: no, I meant to check if the component was left without finish editing. I just followed what we have in YourCompany component: `isEditingNames`, `isEditingType` etc.

File: src/components/Articles/ResolutionDates.vue
Comment (severinbeauvais) [inline src/components/Articles/ResolutionDates.vue:209]: If it makes sense to you, move this (and maybe line 216) to the state-getters module (which is where some other such checks exist).
I'm not saying to change this now... I would say we don't have a complete design for such methods/getters atm so I leave it up to you.

Comment (severinbeauvais) [reply src/components/Articles/ResolutionDates.vue:209]: Oh, I now see you moved some of this out of state-getters. I'm sure you had a good reason. Ignore comment above and carry on :)

Comment (andrepestana-aot) [reply src/components/Articles/ResolutionDates.vue:209]: I don't like the idea of adding a lot of logic to state-getters... actually it was in a state-getter and I refactored it to here.

Comment (andrepestana-aot) [reply src/components/Articles/ResolutionDates.vue:209]: And, yes, I had a good reason. We need to discuss.

File: tests/unit/ResolutionDates.spec.ts
Comment (severinbeauvais) [inline tests/unit/ResolutionDates.spec.ts:288]: Great test!

File: src/components/Articles/Articles.vue
Comment (cameron-freshworks) [reply src/components/Articles/Articles.vue:56]: +1 to Severins comment. 
In this case it makes sense to rename the component event to something more generic, as we KNOW that event belongs to the component, where the local property needs to be specific.

Comment (severinbeauvais) [reply src/components/Articles/Articles.vue:56]: The local variable name is fine, but I think we should have consistent event names.

File: src/components/Articles/ResolutionDates.vue
Comment (cameron-freshworks) [reply src/components/Articles/ResolutionDates.vue:209]: I disagree, i think it makes sense for this getter to be global, as you are now pulling shareStructure information into this component.

I think using getters to compute some value is fine as long as it's specific and clear in its intent. 
My preference is to keep the components as simple as possible (i know it's tough sometimes), but moving this logic into the component only achieves 1 thing imo and that is making in accessible to anywhere else.

Comment (severinbeauvais) [reply src/components/Articles/ResolutionDates.vue:209]: As I said, we don't have a complete design for this :)

What do you (y'all) think of a mixin for methods with store logic?

Comment (andrepestana-aot) [reply src/components/Articles/ResolutionDates.vue:209]: > I disagree, i think it makes sense for this getter to be global, as you are now pulling shareStructure information into this component.
Yes, I agree that this is not ideal dealing with shareStructure here.

> My preference is to keep the components as simple as possible (i know it's tough sometimes), but moving this logic into the component only achieves 1 thing imo and that is making in accessible to anywhere else.
I agree on keeping the components as simple as possible. I think we can find another approach other than using state-getters because when you do it, you are overriding the real state (the state will show is `isValidResolutionDate === true` but the getter may result in false for example)

Creating a mixin is a better alternative imo.

Comment (severinbeauvais) [reply src/components/Articles/ResolutionDates.vue:209]: > I think using getters to compute some value is fine as long as it's specific and clear in its intent.
My preference is to keep the components as simple as possible (i know it's tough sometimes), but moving this logic into the component only achieves 1 thing imo and that is making in accessible to anywhere else.

I strongly agree with this.

However, atm it's a guideline and not a strong design.

**Review by cameron-freshworks — COMMENTED** [review]
I've noticed a few things when running locally. 

The validation is still allowing you to proceed through to Review mode even if your Resolution Date is invalid. 

Also, the border isn't appearing on the invalid resolution date section.

Screen Shot 2021-04-28 at 10 35 11 AM

Screen Shot 2021-04-28 at 10 35 17 AM

File: src/components/Articles/ResolutionDates.vue
Comment (cameron-freshworks) [reply src/components/Articles/ResolutionDates.vue:209]: I don't fully understand how a getter is overriding real* state, where the computed value is merely a representation of what the store values are. 

Mixin's are great for sure, but are purposed for used/extended by components. 
I think we've used getters to compute some values in the past because the easy access to store, aka `state`.

There is always a better way for sure, i'm 100% open to trying anything!

Comment (severinbeauvais) [reply src/components/Articles/ResolutionDates.vue:209]: Andre, why don't you try something that makes sense to you (looks/feels like good design), and we can review it and hopefully follow along with our later work?

**Comment by andrepestana-aot** [comment]
> I've noticed a few things when running locally.
> 
> The validation is allowing you to proceed through to Review mode even if your Resolution Date is invalid.

How do I test the Review mode?

> Also, the border isn't appearing on the invalid resolution date section.
> 
> The invalid text under the Review and certify isn't appearing either.

I see what you're saying but this is before you click on "Review and Certify". When you click on "Review and Certify" it will trigger the component validations, and then, the border and the message will be shown. I thought that this would be the behaviour expected.

What's the behaviour you expect in this case?

File: src/components/Articles/Articles.vue
Comment (andrepestana-aot) [reply src/components/Articles/Articles.vue:56]: >The local variable name is fine, but I think we should have consistent event names.

Do you guys think I should change the name? I followed what we have for other components like in YourCompany.

Comment (severinbeauvais) [reply src/components/Articles/Articles.vue:56]: Yes, please change the event name (not the variable name).

**Comment by cameron-freshworks** [comment]
> > I've noticed a few things when running locally.
> > The validation is allowing you to proceed through to Review mode even if your Resolution Date is invalid.
> 
> How do I test the Review mode?
> 
> > Also, the border isn't appearing on the invalid resolution date section.
> > The invalid text under the Review and certify isn't appearing either.
> 
> I see what you're saying but this is before you click on "Review and Certify". When you click on "Review and Certify" it will trigger the component validations, and then, the border and the message will be shown. I thought that this would be the behaviour expected.
> 
> What's the behaviour you expect in this case? Right away after the shareStructure is added it should show the red border? The message is only shown after clicking on the button.

Review mode is the second screen, once you've made changes you can click Review and Confirm, and if all is valid you can see the review/summary mode. Currently you are able to proceed even if the resolution date section is invalid. 

You are right that you won't see the text under the Review and Certify button until clicked, but in this case it isn't showing even after clicking. 

With Resolution Date you are able to be invalid before clicking review and certify, for example if you Add a share class with Special rights and restrictions and add a DATE, but then remove the date, you will be in an invalid state, which should prevent proceeding. 

Expected behaviour is the Resolution Date section should be invalid if you remove the date required for any special rights or restrictions.

**Comment by andrepestana-aot** [comment]
@cameron-freshworks Thanks for pointing out the validation issue!

I decided to keep the logic in the state-getter for now. We can think of a better approach later. I also fixed issues with the scrolling. I checked all the scrolling and there's still something to be improved for shareStructure because all the scrolling is going to the start of the red border but as shareStructure has a dynamic number of divs, for this one the scroll takes the user to the top of the section.

File: src/components/Articles/Articles.vue
Comment (severinbeauvais) [inline src/components/Articles/Articles.vue:17]: I was hoping the event would simply be called "isEditing" (and same for all components).

(Admittedly I would prefer the event on line 15 to be called "isChanged". But nvm that.)

The "on" naming is usually for event handlers, eg,
`@isEditing="onEditingDoSomething($event)"`

Comment (andrepestana-aot) [reply src/components/Articles/Articles.vue:17]: ok. I can change to `isEditing`. 👍

File: src/components/Articles/Articles.vue
Comment (severinbeauvais) [inline src/components/Articles/Articles.vue:30]: Again, this is really an event telling us if the component is in editing mode, right? So I would use the same verb as above: `@isEditing="..."`

What I'm doing here is taking a step back and looking at things... the Articles component doesn't care about the details of Resolution Dates component. It only cares that Resolution Dates is in edit mode. So that's all I would keep track of. And if we can make the component event handers consistent then that will be helpful for someone to generally handle the code. This project is almost big enough now that I don't understand everything in it, so keep my "immediate concerns" in scope and not thinking about what sub-components do is very helpful.

File: src/components/Articles/Articles.vue
Comment (severinbeauvais) [reply src/components/Articles/Articles.vue:17]: Thanks. That will be cleaner.

Also, other events can be "isValid", "isChanged" (or "hasChanges" or "hasChanged")... yeah we're still working on naming :)

Re: hardest problems in computer science - https://martinfowler.com/bliki/TwoHardThings.html

File: src/components/Articles/Articles.vue
Comment (andrepestana-aot) [reply src/components/Articles/Articles.vue:30]: Yes, I understand what you're saying but I need to know if the subcomponent isEditing to set the invalid state for the resolution-dates component. editMode is another thing related to check if the component is "editable".

Comment (severinbeauvais) [reply src/components/Articles/Articles.vue]: isAdding ?

Comment (andrepestana-aot) [reply src/components/Articles/Articles.vue]: yes, I changed onAddingResolutionDate to isEditing to keep it consistent to the other components.

**Review by severinbeauvais — APPROVED** [review]

File: package.json
Comment (cameron-freshworks) [inline package.json:3]: i believe `main` is now at `6.3` so this might require another bump up

**Comment by cameron-freshworks** [comment]
> @cameron-freshworks Thanks for pointing out the validation issue!
> 
> I decided to keep the logic in the state-getter for now. We can think of a better approach later. I also fixed issues with the scrolling. I checked all the scrolling and there's still something to be improved for shareStructure because all the scrolling is going to the start of the red border but as shareStructure has a dynamic number of divs, for this one the scroll takes the user to the top of the section.

No prob! I ran this locally as well and it all seems to work great 👍 Good job here, it is a tricky bit :)

**Review by cameron-freshworks — APPROVED** [review]
Looks good, great job 👍 

Just check the version # before merge, might need to bump it up and `npm i` to update your package-lock.

File: package.json
Comment (andrepestana-aot) [reply package.json]: 👍

**Comment by andrepestana-aot** [comment]
@cameron-freshworks Please merge if everything is ok. Thanks!

**Pull request merged**
